tests/parse_metadata.py

```diff
@@ -202,10 +204,16 @@ def read_metadata(distribution: str) -> StubMetadata:
     assert isinstance(upstream_repository, (str, type(None)))
     if isinstance(upstream_repository, str):
         parsed_url = urllib.parse.urlsplit(upstream_repository)
-        assert parsed_url.scheme == "https", "URLs in the upstream_repository field should use https"
-        assert not parsed_url.netloc.startswith("www."), "`www.` should be removed from URLs in the upstream_repository field"
-        assert not parsed_url.query
-        assert not parsed_url.fragment
+        assert parsed_url.scheme == "https", f"{distribution}: URLs in the upstream_repository field should use https"
+        assert not parsed_url.netloc.startswith(
+            "www."
+        ), f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
+        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (
+            not parsed_url.query
+        ), f"{distribution}: Query params (`?`) should be removed from URLs in the upstream_repository field"
+        assert (
+            not parsed_url.fragment
+        ), f"{distribution}: Fragments (`#`) should be removed from URLs in the upstream_repository field"
```

Comment:
I can't bear the way black formats these lines, I'm afraid ;)

```suggestion
        no_www_please = (
            f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
        )
        assert not parsed_url.netloc.startswith("www."), no_www_please
        no_query_params_please = (
            f"{distribution}: Query params (`?`) should be removed from URLs in the upstream_repository field"
        )
        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (not parsed_url.query), no_query_params_please
        no_fragments_please = f"{distribution}: Fragments (`#`) should be removed from URLs in the upstream_repository field"
        assert not parsed_url.fragment, no_fragments_please
```

Reply:
This is one of my big pet peaves with black too. I'll go with your changes.